lib/location/method_wifi.c

```diff
@@ -0,0 +1,307 @@
+	"At least one WiFi positioning service must be enabled");
+
+extern location_event_handler_t event_handler;
+extern struct loc_event_data current_event_data;
+
+struct k_work_args {
+	struct k_work work_item;
+	struct loc_wifi_config wifi_config;
+	int64_t starting_uptime_ms;
+
+};
+
+static struct net_if *wifi_iface;
+
+static struct k_work_args method_wifi_start_work;
+
+static bool running;
+
+static uint32_t current_scan_result_count;
+static uint32_t latest_scan_result_count;
+
+struct method_wifi_scan_result {
+	char mac_addr_str[WIFI_MAC_ADDR_STR_LEN + 1];
+	char ssid_str[WIFI_SSID_MAX_LEN + 1];
+	uint8_t channel;
+	int8_t rssi;
+};
+
+static struct method_wifi_scan_result
+	latest_scan_results[CONFIG_LOCATION_METHOD_WIFI_SCANNING_RESULTS_MAX_CNT];
+static K_SEM_DEFINE(wifi_scanning_ready, 0, 1);
+
+/******************************************************************************/
+
+static int method_wifi_scanning_start(void)
+{
+	int ret;
+
+	LOG_DBG("Triggering start of WiFi scanning");
+
+	latest_scan_result_count = 0;
+	current_scan_result_count = 0;
+
+	if (wifi_iface != NULL) {
+		ret = net_mgmt(NET_REQUEST_WIFI_SCAN, wifi_iface, NULL, 0);
+		if (ret) {
+			LOG_ERR("Failed to initiate WiFi scanning: %d", ret);
+			ret = -1;
+		}
+	} else {
+		LOG_ERR("Failed to initiate WiFi scanning: no wifi device");
+		ret = -1;
+	}
+	return ret;
+}
+
+/******************************************************************************/
+
+static void method_wifi_handle_wifi_scan_result(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_scan_result *entry = (const struct wifi_scan_result *)cb->info;
+	struct method_wifi_scan_result *current;
+
+	current_scan_result_count++;
+	current = &latest_scan_results[current_scan_result_count - 1];
+
+	if (current_scan_result_count <= CONFIG_LOCATION_METHOD_WIFI_SCANNING_RESULTS_MAX_CNT) {
+		sprintf(current->mac_addr_str,
+			"%02x:%02x:%02x:%02x:%02x:%02x",
+			entry->mac[0], entry->mac[1], entry->mac[2],
+			entry->mac[3], entry->mac[4], entry->mac[5]);
+		snprintf(current->ssid_str, entry->ssid_length + 1, "%s", entry->ssid);
+
+		current->channel = entry->channel;
+		current->rssi = entry->rssi;
+
+		LOG_DBG("scan result #%d stored: ssid %s, mac address: %s, channel %d,",
+			current_scan_result_count,
+			log_strdup(current->ssid_str),
+			log_strdup(current->mac_addr_str),
+			current->channel);
+	} else {
+		LOG_WRN("Scanning result (mac %02x:%02x:%02x:%02x:%02x:%02x) "
+			"did not fit to result buffer - dropping it",
+				entry->mac[0], entry->mac[1], entry->mac[2],
+				entry->mac[3], entry->mac[4], entry->mac[5]);
+	}
+}
+
+static void method_wifi_handle_wifi_scan_done(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_status *status = (const struct wifi_status *)cb->info;
+
+	if (status->status) {
+		LOG_WRN("WiFi scan request failed (%d).", status->status);
+	} else {
+		LOG_INF("Scan request done.");
+	}
+	k_sem_give(&wifi_scanning_ready);
```

Comment:
Would it be safer to give the semaphore after setting `latest_scan_result_count`, which is used in the function that's waiting for this semaphore?

Reply:
yes. changed.